modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheGateway.java

```diff
@@ -352,7 +355,12 @@ public void onStopped() {
      * @throws IgniteException - in case of atomic operation inside transaction without permission.
      */
     private void checkAtomicOpsInTx(CacheOperationContext opCtx) throws IgniteException {
-        if (ctx.atomic() && !opCtx.allowedAtomicOpsInTx()) {
+        CacheConfiguration cfg = ctx.config();
+
+        if (cfg == null)
```

Comment:
Why do we rely on cfg == null? Looks very fragile.

Reply:
Unfortunately, this method is called from `GridCacheGateway#enter(CacheOperationContext)` before the `readLock` is acquired, and therefore there is no guarantee that the cache context was not cleaned. Perhaps, it makes sense to move this check under the mentioned lock.